plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageServiceImpl.java

```diff
@@ -34,167 +33,137 @@
 import com.microsoft.azure.storage.blob.DeleteSnapshotsOption;
 import com.microsoft.azure.storage.blob.ListBlobItem;
 import org.apache.logging.log4j.message.ParameterizedMessage;
-import org.apache.logging.log4j.util.Supplier;
 import org.elasticsearch.common.blobstore.BlobMetaData;
 import org.elasticsearch.common.blobstore.support.PlainBlobMetaData;
 import org.elasticsearch.common.collect.MapBuilder;
+import org.elasticsearch.common.collect.Tuple;
 import org.elasticsearch.common.component.AbstractComponent;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.settings.SettingsException;
 import org.elasticsearch.repositories.RepositoryException;
 
 import java.io.InputStream;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.security.InvalidKeyException;
 import java.util.EnumSet;
-import java.util.HashMap;
 import java.util.Map;
+import java.util.function.Supplier;
 
-public class AzureStorageServiceImpl extends AbstractComponent implements AzureStorageService {
+import static java.util.Collections.emptyMap;
 
-    final Map<String, AzureStorageSettings> storageSettings;
+public class AzureStorageServiceImpl extends AbstractComponent implements AzureStorageService {
 
-    final Map<String, CloudBlobClient> clients = new HashMap<>();
+    // 'package' for testing
+    volatile Map<String, AzureStorageSettings> storageSettings = emptyMap();
 
-    public AzureStorageServiceImpl(Settings settings, Map<String, AzureStorageSettings> storageSettings) {
+    public AzureStorageServiceImpl(Settings settings) {
         super(settings);
-
-        this.storageSettings = storageSettings;
-
-        if (storageSettings.isEmpty()) {
-            // If someone did not register any settings, they basically can't use the plugin
+        // eagerly load client settings so that secure settings are read
+        final Map<String, AzureStorageSettings> clientsSettings = AzureStorageSettings.load(settings);
```

Comment:
I have moved the validation inside the `AzureStorageSettings#load` method, where settings are constructed.

I have contemplated this duplication before commiting it. I have tried to keep settings validation closer to the "use-point" instead of "construction point" since, in principle, they might change in other layers (which is not the case here). This was the closest I could get, with only duplication, because of the override location mode mechanism which temporarily updates with invalid settings (empty).